torch/csrc/jit/ir.cpp

```diff
@@ -875,6 +875,197 @@ Node* Node::insertAfter(Node * n) {
+    for (const auto output : n->outputs()) {
+      for (const auto& use : output->uses()) {
+        if (use.user->owningBlock() == n->owningBlock()) {
+          users.insert(use.user);
+        } else {
+          // This user is in a sub-block. Traverse the blockchain upward until
+          // we arrive at a node that shares a block with `this`
+          auto curNode = use.user;
+          while (curNode->owningBlock() != n->owningBlock()) {
+            curNode = curNode->owningBlock()->owningNode();
+          }
+          users.insert(curNode);
+        }
+      }
+    }
+
+    return users;
+  }
+
+  std::list<Node*> nodes_;
+  std::unordered_set<Node*> users_;
+};
+} // namespace
+
+// Try to move `this` before/after `movePoint` while preserving value
+// dependencies. Returns false iff such a move could not be made
+//
+// The basic approach is: have a "working set" that we are moving forward, one
+// node at a time. When we can't move past a node (because it depends on the
+// working set), then add it to the working set and keep moving until we hit
+// `moveAfter`.
+bool Node::tryMove(Node* movePoint, MoveSide moveSide) {
+  JIT_ASSERT(this->inBlockList() && movePoint->inBlockList());
+  JIT_ASSERT(this->owningBlock() == movePoint->owningBlock());
+  JIT_ASSERT(this != movePoint);
+
+  // 1. Move from `this` toward movePoint, building up the working set of
+  // dependencies
+  WorkingSet workingSet(this);
+
+  int direction;
+  if (this->isAfter(movePoint)) {
+    direction = kPrevDirection;
+  } else {
+    direction = kNextDirection;
+  }
+
+  auto curNode = this->next_in_graph[direction];
+  // Move forward one node at a time
+  while (curNode != movePoint) {
+    if (workingSet.dependsOn(curNode)) {
+      // If we can't move past this node, add it to the working set
+      workingSet.add(curNode);
+    }
+    curNode = curNode->next_in_graph[direction];
+  }
+
+  // 2. Decide whether we can move it all to `movePoint`.
+
+  // Say we are moving directly before movePoint and `this` starts before
+  // movePoint in the graph. The move looks like
+  //
+  //  `this`              `this`           |
+  //  <dependencies>  ->  `movePoint`      | `this` and deps are split
+  //  `movePoint`         <dependencies>   |
+  //
+  // Contrast with the case where `this` starts AFTER movePoint:
+  //
+  //  `movePoint`         <dependencies>   |
+  //  <dependencies>  ->  `this`           | `this` and deps are together
+  //  `this`              `movePoint`      |
+  //
+  // In the first case, we need to split `this` off from its dependencies, so we
+  // can move the dependencies below `movePoint` and keep `this` above.
+  const bool splitThisAndDeps =
+      (moveSide == MoveSide::BEFORE && this->isBefore(movePoint)) ||
+      (moveSide == MoveSide::AFTER && this->isAfter(movePoint));
+
+  if (splitThisAndDeps) {
+    // remove `this` from dependencies to be moved past `movePoint`
+    workingSet.eraseMover();
```

Comment:
eraseMover doesn't accurately update the uses_ set in workingSet. So the dependsOn query below can still contain false uses. I think this prevents some legitimate moves. If uses_ was a map, you can keep count of how may times it appeared as a use of the set, and then eraseMover can decrement and remove if the count is 0.

Reply:
whoops, you're right. I'll fix